In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
pd.set_option('display.max_columns', None)
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [29]:
# LU look up file for health codes
## refer to "inpatient_data_preprocessing_.ipynb"
LU_ICD = pd.read_csv('/content/drive/MyDrive/Data/icd10/ICD10Diagnosis.csv')
LU_ICD.rename(columns={'Code': 'ICD_DIAG_CD', 'Description': 'ICD_Description'}, inplace=True)

outpatient = pd.read_csv('/content/drive/MyDrive/Data/Input/outpatient.csv', sep='|')
col_outpatient = pd.DataFrame(outpatient.dtypes, columns=['type'])
#col_outpatient

<ipython-input-29-dada294be7ed>:6: DtypeWarning: Columns (9,55,56,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117) have mixed types. Specify dtype option on import or set low_memory=False.
  outpatient = pd.read_csv('/content/drive/MyDrive/Data/Input/outpatient.csv', sep='|')


In [30]:
# convert columns to appropriate data types
outpatient['BENE_ID'] = outpatient['BENE_ID'].astype(str)
outpatient['CLM_ID'] = outpatient['CLM_ID'].astype(str)
outpatient['PTNT_DSCHRG_STUS_CD'] = outpatient['PTNT_DSCHRG_STUS_CD'].astype(str)

# dates columns
outpatient['CLM_FROM_DT'] = pd.to_datetime(outpatient['CLM_FROM_DT'], format='%d-%b-%Y')
outpatient['CLM_THRU_DT'] = pd.to_datetime(outpatient['CLM_THRU_DT'], format='%d-%b-%Y')
outpatient['YR'] = outpatient['CLM_FROM_DT'].dt.year

# drop and select columns
columns_to_drop = [col for col in outpatient.columns if col.endswith('UPIN') or col.startswith('ICD_DGNS_E_CD') or col.startswith('PRCDR_DT')]
outpatient = outpatient.drop(columns=columns_to_drop)
outpatient

BENE_ID           CLM_ID NCH_NEAR_LINE_REC_IDENT_CD  \
0       -10000010254618  -10000930037832                          W   
1       -10000010254618  -10000930037833                          W   
2       -10000010254618  -10000930037834                          W   
3       -10000010254618  -10000930037835                          W   
4       -10000010254618  -10000930037836                          W   
...                 ...              ...                        ...   
575087  -10000010288007  -10000931486350                          W   
575088  -10000010288007  -10000931486351                          W   
575089  -10000010288007  -10000931486352                          W   
575090  -10000010288007  -10000931486353                          W   
575091  -10000010288007  -10000931486354                          W   

        NCH_CLM_TYPE_CD CLM_FROM_DT CLM_THRU_DT NCH_WKLY_PROC_DT  \
0                    40  2015-03-16  2015-03-16      20-Mar-2015   
1                    40  2015-04-27  2015-04-27      01-May-2015   
2                    40  2015-05-23  2015-05-23      29-May-2015   
3                    40  2015-06-01  2015-06-01      05-Jun-2015   
4                    40  2015-06-14  2015-06-14      19-Jun-2015   
...                 ...         ...         ...              ...   
575087               40  2020-11-05  2020-11-05      06-Nov-2020   
575088               40  2021-05-11  2021-05-11      14-May-2021   
575089               40  2021-06-08  2021-06-08      11-Jun-2021   
575090               40  2022-06-11  2022-06-11      17-Jun-2022   
575091               40  2022-09-11  2022-09-11      16-Sep-2022   

        FI_CLM_PROC_DT  CLAIM_QUERY_CODE PRVDR_NUM  CLM_FAC_TYPE_CD  \
0                  NaN                 3    01S023                1   
1                  NaN                 3    01S023                1   
2                  NaN                 3    01S023                1   
3                  NaN                 3    01S023                1   
4                  NaN                 3    01S023                1   
...                ...               ...       ...              ...   
575087             NaN                 3    050789                1   
575088             NaN                 3    050789                1   
575089             NaN                 3    050789                1   
575090             NaN                 3    050789                1   
575091             NaN                 3    050789                1   

        CLM_SRVC_CLSFCTN_TYPE_CD  CLM_FREQ_CD  FI_NUM CLM_MDCR_NON_PMT_RSN_CD  \
0                              3            1     NaN                           
1                              3            1     NaN                           
2                              3            1     NaN                           
3                              3            1     NaN                           
4                              3            1     NaN                           
...                          ...          ...     ...                     ...   
575087                         3            1     NaN                           
575088                         3            1     NaN                           
575089                         3            1     NaN                           
575090                         3            1     NaN                           
575091                         3            1     NaN                           

        CLM_PMT_AMT  NCH_PRMRY_PYR_CLM_PD_AMT NCH_PRMRY_PYR_CD  \
0          10224.89                      0.00                    
1          11936.39                      0.00                    
2           8674.15                      0.00                    
3           1171.57                      0.00                    
4          11327.22                      0.00                    
...             ...                       ...              ...   
575087       326.65                      0.00             

In [31]:
outpatient['REV_CNTR'].value_counts()

REV_CNTR
1      574480
780       317
789       295
Name: count, dtype: int64

* `rev_center` (revenue center) of 450 means ER
* No ER claims exist in this data (but expect lots of 450s in the inpatient file, like we saw based on `ER_flag` in "inpatient_data_preprocessing_.ipynb")
* `rev_center` of 780 and 789 ralates with telemedicine, so we can ignore...

In [33]:
outpatient = outpatient[outpatient['CLM_LINE_NUM'] == 1]

#compute LOS (NB: usually 1 for all outpatient data, but could be more than 1 day for patients under observation/recovery)
outpatient['LOS'] = (outpatient['CLM_THRU_DT'] - outpatient['CLM_FROM_DT']).dt.days + 1
outpatient['LOS'].value_counts()

LOS
1     371460
2      31087
3         81
8         22
4          2
48         1
Name: count, dtype: int64

In [34]:
diagnosis = outpatient[['BENE_ID', 'YR', 'PRNCPAL_DGNS_CD'] + [col for col in outpatient.columns if col.startswith('ICD_DGNS_CD')]]

# wide to long format data, and remove rows where ICD_DIAG_CD is not populated
diagnosis = pd.melt(diagnosis, id_vars=['BENE_ID', 'YR'], var_name='ICD_DIAG_COL', value_name='ICD_DIAG_CD')
diagnosis = diagnosis[diagnosis['ICD_DIAG_CD'].notna() & (diagnosis['ICD_DIAG_CD'] != '')]

# drop the 'ICD_DIAG_COL', then remove duplicates
diagnosis = diagnosis.drop(columns=['ICD_DIAG_COL'])
diagnosis = diagnosis.drop_duplicates()

diagnosis[diagnosis['BENE_ID'] == '-10000010254618'] #randomly viewing a beneficiary

BENE_ID    YR ICD_DIAG_CD
0        -10000010254618  2015       R4689
14       -10000010254618  2016       R4689
22       -10000010254618  2016       B9789
33       -10000010254618  2017       B9789
48       -10000010254618  2018       B9789
...                  ...   ...         ...
4831900  -10000010254618  2021        E848
4831908  -10000010254618  2022        E848
4831909  -10000010254618  2023        N469
5234546  -10000010254618  2020        N469
5234558  -10000010254618  2022        N469

[143 rows x 3 columns]

In [35]:
#  merge with ICD data
diagnosis = pd.merge(diagnosis, LU_ICD, on='ICD_DIAG_CD', how='left')
num_diagnosis = diagnosis.groupby(['BENE_ID', 'YR']).size().reset_index(name='NUM_DIAG')
num_diagnosis.head(5)

BENE_ID    YR  NUM_DIAG
0  -10000010254618  2015        13
1  -10000010254618  2016        14
2  -10000010254618  2017        15
3  -10000010254618  2018        13
4  -10000010254618  2019        18

In [36]:
outpatient_encounters = outpatient[['BENE_ID', 'CLM_ID', 'CLM_FROM_DT', 'CLM_THRU_DT', 'YR', #'LOS',
                                  'PRNCPAL_DGNS_CD', 'PTNT_DSCHRG_STUS_CD', 'CLM_TOT_CHRG_AMT']]

# perform left joins
outpatient_encounters = outpatient_encounters.merge(num_diagnosis, on=['BENE_ID', 'YR'], how='left')
outpatient_encounters = outpatient_encounters.merge(LU_ICD, left_on='PRNCPAL_DGNS_CD', right_on='ICD_DIAG_CD',
                                                  how='left').drop(columns=['ICD_DIAG_CD'])
outpatient_encounters

BENE_ID           CLM_ID CLM_FROM_DT CLM_THRU_DT    YR  \
0       -10000010254618  -10000930037832  2015-03-16  2015-03-16  2015   
1       -10000010254618  -10000930037833  2015-04-27  2015-04-27  2015   
2       -10000010254618  -10000930037834  2015-05-23  2015-05-23  2015   
3       -10000010254618  -10000930037835  2015-06-01  2015-06-01  2015   
4       -10000010254618  -10000930037836  2015-06-14  2015-06-14  2015   
...                 ...              ...         ...         ...   ...   
402648  -10000010288007  -10000931486350  2020-11-05  2020-11-05  2020   
402649  -10000010288007  -10000931486351  2021-05-11  2021-05-11  2021   
402650  -10000010288007  -10000931486352  2021-06-08  2021-06-08  2021   
402651  -10000010288007  -10000931486353  2022-06-11  2022-06-11  2022   
402652  -10000010288007  -10000931486354  2022-09-11  2022-09-11  2022   

       PRNCPAL_DGNS_CD PTNT_DSCHRG_STUS_CD  CLM_TOT_CHRG_AMT  NUM_DIAG  \
0                R4689                   1          10224.89        13   
1                R4689                   1          11936.39        13   
2                R4689                   1           8674.15        13   
3                R4689                   1           1171.57        13   
4                R4689                   1          11327.22        13   
...                ...                 ...               ...       ...   
402648            Z940                   1            326.65        21   
402649            Z940                   1            272.70        17   
402650            Z940                   1            272.70        17   
402651          S5290X                   1             82.02        18   
402652          S5290X                   1          15115.26        18   

                                          ICD_Description  
0       Other symptoms and signs involving appearance ...  
1       Other symptoms and signs involving appearance ...  
2       Other symptoms and signs involving appearance ...  
3       Other symptoms and signs involving appearance ...  
4       Other symptoms and signs involving appearance ...  
...                                                   ...  
402648                           Kidney transplant status  
402649                           Kidney transplant status  
402650                           Kidney transplant status  
402651  A Unspecified fracture of unspecified forearm,...  
402652  A Unspecified fracture of unspecified forearm,...  

[402653 rows x 10 columns]

In [37]:
outpatient_encounters.to_csv('/content/drive/MyDrive/Data/Output/outpatient_encounters.csv', index=False)
diagnosis.to_csv('/content/drive/MyDrive/Data/Output/outpatient_diagnosis.csv', index=False)